param.go

```diff
@@ -0,0 +1,187 @@
+package dig
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// The param interface represents a dependency for a constructor.
+//
+// The following implementations exist:
+//  paramList     All arguments of the constructor.
+//  paramSimple   An explicitly requested type.
+//  paramObject   dig.In struct where each field in the struct can be another
+//                param.
+type (
+	param interface {
+		param()
+
+		// Human-friendly name for this param type. We use this exclusively in
+		// error messages.
+		ParamTypeName() string
```

Comment:
`param()` is private to conform with the pattern of declaring an unexported
method on an interface which ensures that all implementations of that
interface are in that same package.

The rest are public because of the reason stated above.